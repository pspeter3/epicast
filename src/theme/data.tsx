import * as React from "react";

export const Meter: React.SFC<{ active?: boolean; value: number }> = ({ active, value }) => {
    const classNames = ["data-meter"];
    if (active) {
        classNames.push("active");
    }
    return <meter className={classNames.join(" ")} value={value} />;
};

export const Percentage: React.SFC<{ value: number }> = ({ value }) => {
    return <p className="data-percentage">{Math.round(value * 100)}%</p>;
};
