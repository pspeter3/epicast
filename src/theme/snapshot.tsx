import * as React from "react";
import TestRenderer from "react-test-renderer";

export const snapshot = (name: string, SFC: React.SFC<{}>): void => {
    describe(name, () => {
        it(`should match snapshot`, () => {
            const component = TestRenderer.create(<SFC>Snapshot</SFC>);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
};
