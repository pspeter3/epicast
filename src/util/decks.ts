import { Deck } from "../core/types";
import { size } from "./stacks";

export const unique = (deck: Deck): string[] =>
    Object.keys(
        deck.reduce(
            (set, stack) => {
                return Object.keys(stack).reduce((collection, key) => {
                    collection[key] = null;
                    return collection;
                }, set);
            },
            {} as Record<string, null>,
        ),
    );

export const total = (deck: Deck): number => deck.reduce((sum, stack) => sum + size(stack), 0);
