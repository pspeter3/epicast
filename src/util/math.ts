export const factorial = (n: number): number => {
    if (n < 0) {
        throw new Error(`Invalid range: ${n} >= 0`);
    }
    if (n === 0) {
        return 1;
    }
    return n * factorial(n - 1);
};

export const combinations = (n: number, k: number): number => {
    if (k > n) {
        throw new Error(`Invalid range: ${n} >= ${k}`);
    }
    return factorial(n) / (factorial(k) * factorial(n - k));
};

export const hypergeometric = (N: number, K: number, n: number, k: number): number => {
    const min = Math.max(0, n + K - N);
    const max = Math.min(K, n);
    if (k < min || k > max) {
        throw new Error(`Invalid range: ${max} >= ${k} >= ${min}`);
    }
    return combinations(K, k) * combinations(N - K, n - k) / combinations(N, n);
};
