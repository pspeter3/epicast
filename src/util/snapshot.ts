import { shallow } from "enzyme";
import * as React from "react";

export type Factory = () => React.ReactElement<any>;

export const snapshot = (factory: Factory): (() => void) => {
    return () => {
        expect(shallow(factory())).toMatchSnapshot();
    };
};

export const snapshotSuite = (name: string, factory: Factory): void => {
    describe(name, () => {
        it("should match snapshot", snapshot(factory));
    });
};
