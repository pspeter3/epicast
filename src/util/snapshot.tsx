import * as React from "react";
import TestRenderer from "react-test-renderer";

export type Factory = () => React.ReactElement<any>;

export const snapshot = (factory: Factory): (() => void) => {
    return () => {
        const component = TestRenderer.create(factory());
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    };
};

export const snapshotSuite = (name: string, factory: Factory): void => {
    describe(name, () => {
        it("should match snapshot", snapshot(factory));
    });
};
