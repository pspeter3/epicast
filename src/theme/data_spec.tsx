import * as React from "react";
import { snapshot } from "../util/snapshot";
import { Meter, Percentage } from "./data";

describe("Data", () => {
    describe("Meter", () => {
        it("should match snapshot", snapshot(() => <Meter value={0.5} />));
        it("should match active snapshot", snapshot(() => <Meter active value={0.5} />));
    });

    describe("Percentage", () => {
        it("should match snapshot", snapshot(() => <Percentage value={0.5} />));
        it("should match rounded snapshot", snapshot(() => <Percentage value={0.502} />));
    });
});
