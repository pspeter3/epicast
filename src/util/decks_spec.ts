import { total, unique } from "./decks";

const PRIMARY = "San Francisco";
const SECONDARY = "Los Angeles";

describe("decks", () => {
    describe("unique", () => {
        it("should return the unique cards", () => {
            expect(
                unique([
                    { [PRIMARY]: 1 },
                    {
                        [PRIMARY]: 2,
                        [SECONDARY]: 1,
                    },
                ]),
            ).toEqual([PRIMARY, SECONDARY]);
        });
    });

    describe("total", () => {
        it("should return the exact count", () => {
            expect(
                total([
                    { [PRIMARY]: 1 },
                    {
                        [PRIMARY]: 2,
                        [SECONDARY]: 1,
                    },
                ]),
            ).toBe(4);
        });
    });
});
