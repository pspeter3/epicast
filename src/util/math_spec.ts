import { combinations, factorial, hypergeometric } from "./math";

describe("math", () => {
    describe("factorial", () => {
        it("should throw if less than zero", () => {
            expect(() => {
                factorial(-1);
            }).toThrowErrorMatchingSnapshot();
        });

        it("should handle 0", () => {
            expect(factorial(0)).toBe(1);
        });

        it("should handle normal numbers", () => {
            expect(factorial(3)).toBe(6);
        });
    });

    describe("combinations", () => {
        it("should throw if k is less than n", () => {
            expect(() => {
                combinations(3, 5);
            }).toThrowErrorMatchingSnapshot();
        });

        it("should handle normal numbers", () => {
            expect(combinations(3, 2)).toBe(3);
        });
    });

    describe("hypergeometric", () => {
        it("should throw if K is greater than N", () => {
            expect(() => {
                hypergeometric(0, 1, 1, 1);
            }).toThrowErrorMatchingSnapshot();
        });

        it("should throw if k is greater than n", () => {
            expect(() => {
                hypergeometric(2, 2, 1, 2);
            }).toThrowErrorMatchingSnapshot();
        });

        it("should throw if k is greater than K", () => {
            expect(() => {
                hypergeometric(1, 0, 1, 1);
            }).toThrowErrorMatchingSnapshot();
        });

        it("should throw if n is greater than N", () => {
            expect(() => {
                hypergeometric(1, 1, 2, 1);
            }).toThrowErrorMatchingSnapshot();
        });

        it("should return the probability", () => {
            expect(hypergeometric(52, 4, 1, 1)).toBe(1 / 13);
        });
    });
});
