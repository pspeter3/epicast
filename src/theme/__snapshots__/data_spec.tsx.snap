// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Meter should match active snapshot 1`] = `
<meter
  className="data-meter active"
  value={0.5}
/>
`;

exports[`Data Meter should match snapshot 1`] = `
<meter
  className="data-meter"
  value={0.5}
/>
`;

exports[`Data Percentage should match rounded snapshot 1`] = `
<p
  className="data-percentage"
>
  50
  %
</p>
`;

exports[`Data Percentage should match snapshot 1`] = `
<p
  className="data-percentage"
>
  50
  %
</p>
`;
